util/topsql/reporter/pubsub.go

```diff
@@ -0,0 +1,268 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package reporter
+
+import (
+	"context"
+	"time"
+
+	"github.com/pingcap/tidb/util"
+	"github.com/pingcap/tidb/util/logutil"
+	"github.com/pingcap/tipb/go-tipb"
+	"go.uber.org/atomic"
+	"go.uber.org/zap"
+)
+
+// TopSQLPubSubService implements tipb.TopSQLPubSubServer.
+//
+// If a client subscribes to TopSQL records, the TopSQLPubSubService is responsible
+// for registering an associated DataSink to the reporter. Then the reporter sends
+// data to the client periodically.
+type TopSQLPubSubService struct {
+	dataSinkRegHandle DataSinkRegHandle
+}
+
+// NewTopSQLPubSubService creates a new TopSQLPubSubService.
+func NewTopSQLPubSubService(
+	dataSinkRegHandle DataSinkRegHandle,
+) *TopSQLPubSubService {
+	return &TopSQLPubSubService{
+		dataSinkRegHandle: dataSinkRegHandle,
+	}
+}
+
+var _ tipb.TopSQLPubSubServer = &TopSQLPubSubService{}
+
+// Subscribe registers dataSinks to the reporter and redirects data received from reporter
+// to subscribers associated with those dataSinks.
+func (t *TopSQLPubSubService) Subscribe(
+	_ *tipb.TopSQLSubRequest,
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) error {
+	ds := newPubSubDataSink(stream)
+
+	if err := t.dataSinkRegHandle.Register(ds); err != nil {
+		return err
+	}
+
+	ds.run()
+	return nil
+}
+
+type pubSubDataSink struct {
+	stream     tipb.TopSQLPubSub_SubscribeServer
+	sendTaskCh chan sendTask
+	isDown     *atomic.Bool
+}
+
+func newPubSubDataSink(
+	stream tipb.TopSQLPubSub_SubscribeServer,
+) *pubSubDataSink {
+	return &pubSubDataSink{
+		stream:     stream,
+		sendTaskCh: make(chan sendTask),
+		isDown:     atomic.NewBool(false),
+	}
+}
+
+var _ DataSink = &pubSubDataSink{}
+
+func (s *pubSubDataSink) Send(data reportData, deadline time.Time) {
+	if s.IsDown() {
+		return
+	}
+
+	select {
+	case s.sendTaskCh <- sendTask{data: data, deadline: deadline}:
+		// sent successfully
+	default:
+		ignoreReportChannelFullCounter.Inc()
+		logutil.BgLogger().Warn("[top-sql] report channel is full")
+	}
+}
+
+func (s *pubSubDataSink) IsPaused() bool {
+	return false
+}
+
+func (s *pubSubDataSink) IsDown() bool {
+	return s.isDown.Load()
+}
+
+func (s *pubSubDataSink) Close() {
+	close(s.sendTaskCh)
+}
+
+func (s *pubSubDataSink) run() {
+	defer s.isDown.Store(true)
+
+	for task := range s.sendTaskCh {
+		ctx, cancel := context.WithDeadline(context.Background(), task.deadline)
+		start := time.Now()
+
+		var err error
+		doneCh := make(chan struct{})
+		go util.WithRecovery(func() {
+			defer func() {
+				doneCh <- struct{}{}
+				cancel()
+			}()
+			err = s.doSend(ctx, task.data)
```

Comment:
Maybe you don't need `doneCh` at all. Calling `cancel()` (which means finished) can be distinguished with a timeout ( by checking the error content ) .